dataset/src/main/scala/frameless/TypedDataset.scala

```diff
@@ -67,6 +83,52 @@ class TypedDataset[T](
     ): GroupedByManyOps[T, K, Out0] = new GroupedByManyOps[T, K, Out0](self, groupedBy)
   }
 
+  def join[A: TypedEncoder, B](
+    right: TypedDataset[A],
+    leftCol: TypedColumn[T, B],
```

Comment:
I don't have a very precise idea, but we could probably have a dedicated "expression" DSL for joins and selects similar to the Spark one, but with all the additional checks we can provide.